python/lsst/meas/astrom/pessimistic_pattern_matcher_b_3D.py

```diff
@@ -314,31 +316,55 @@ def match(self, source_array, n_check, n_match, n_agree,
-
             cut_ids = match_sources_struct.match_ids[
                 match_sources_struct.distances_rad < max_dist_rad]
+
             n_matched = len(cut_ids)
+            # Check clipped distances. The minimum value here
+            # prevents over convergance on perfect test data.
+            clipped_max_dist = np.max((
+                sigmaclip(match_sources_struct.distances_rad,
+                          low=100,
+                          high=2)[-1],
+                1e-16))
+            n_matched_clipped = np.sum(
+                match_sources_struct.distances_rad < clipped_max_dist)
+
+            if n_matched < min_matches or n_matched_clipped < min_matches:
+                continue
 
-            # Check that we have enough matches.
-            if n_matched >= min_matches:
-                # The shift/rotation matrix returned by
-                # ``_construct_pattern_and_shift_rot_matrix``, above, was
-                # based on only six points. Here, we refine that result by
-                # using all of the good matches from the “final verification”
-                # step, above. This will produce a more consistent result.
-                fit_shift_rot_matrix = least_squares(
-                    _rotation_matrix_chi_sq,
-                    x0=shift_rot_matrix.flatten(),
-                    args=(source_array[cut_ids[:, 0], :3],
-                          self._reference_array[cut_ids[:, 1], :3],
-                          max_dist_rad)
-                ).x.reshape((3, 3))
-                match_sources_struct = self._match_sources(
-                    source_array[:, :3], fit_shift_rot_matrix)
+            # The shift/rotation matrix returned by
+            # ``_construct_pattern_and_shift_rot_matrix``, above, was
+            # based on only six points. Here, we refine that result by
+            # using all of the good matches from the “final verification”
+            # step, above. This will produce a more consistent result.
+            fit_shift_rot_matrix = least_squares(
+                _rotation_matrix_chi_sq,
+                x0=shift_rot_matrix.flatten(),
+                args=(source_array[cut_ids[:, 0], :3],
+                      self._reference_array[cut_ids[:, 1], :3],
+                      max_dist_rad)
+            ).x.reshape((3, 3))
+            match_sources_struct = self._match_sources(
+                source_array[:, :3], fit_shift_rot_matrix)
+
+            # Double check the match distances to make sure enough matches
+            # survive still.
+            n_matched = np.sum(
+                match_sources_struct.distances_rad < max_dist_rad)
+            clipped_max_dist = np.max((
+                sigmaclip(match_sources_struct.distances_rad,
+                          low=100,
+                          high=2)[-1],
+                1e-16))
+            n_matched_clipped = np.sum(
+                match_sources_struct.distances_rad < clipped_max_dist)
```

Comment:
There seems to be repeated block of code between lines 324-336 and 354-365. This would be better if it was given some structure, e.g. pulled out into a function. These two blocks also repeat usage of the same variable names, which is both confusing and risky.